covenant/covenant.go

```diff
@@ -215,61 +187,49 @@ func (ce *CovenantEmulator) AddCovenantSignatures(btcDels []*types.Delegation) (
-		// record metrics
-		startSignTime := time.Now()
-		metricsTimeKeeper.SetPreviousSignStart(&startSignTime)
-
-		covenantPrivKey, err := ce.getPrivKey()
+		// 8. Generate Signing Request
+		// Finality providers encryption keys
+		// pk script paths for Slash, unbond and unbonding slashing
+		fpsEncKeys, err := fpEncKeysFromDel(btcDel)
 		if err != nil {
-			return nil, fmt.Errorf("failed to get Covenant private key: %w", err)
+			ce.logger.Error("failed to encript the finality provider keys of the btc delegation", zap.String("staker_pk", stakerPkHex), zap.Error(err))
+			continue
 		}
 
-		slashSigs, unbondingSig, err := signSlashAndUnbondSignatures(
-			btcDel,
-			stakingTx,
-			slashingTx,
-			unbondingTx,
-			covenantPrivKey,
-			params,
-			&ce.config.BTCNetParams,
-		)
+		slashingPkScriptPath, stakingTxUnbondingPkScriptPath, unbondingTxSlashingPkScriptPath, err := pkScriptPaths(btcDel, params, &ce.config.BTCNetParams, unbondingTx)
 		if err != nil {
-			ce.logger.Error("failed to sign signatures or unbonding signature", zap.Error(err))
+			ce.logger.Error("failed to generate pk script path", zap.Error(err))
 			continue
 		}
 
-		// 7. sign covenant slash unbonding signatures
-		slashUnbondingSigs, err := signSlashUnbondingSignatures(
-			btcDel,
-			unbondingTx,
-			slashUnbondingTx,
-			covenantPrivKey,
-			params,
-			&ce.config.BTCNetParams,
-		)
+		// 9. sign covenant transactions
+		resp, err := ce.SignTransactions(SigningRequest{
+			StakingTx:                       stakingTx,
+			SlashingTx:                      slashingTx,
+			UnbondingTx:                     unbondingTx,
+			SlashUnbondingTx:                slashUnbondingTx,
+			StakingOutputIdx:                btcDel.StakingOutputIdx,
+			SlashingPkScriptPath:            slashingPkScriptPath,
+			StakingTxUnbondingPkScriptPath:  stakingTxUnbondingPkScriptPath,
+			UnbondingTxSlashingPkScriptPath: unbondingTxSlashingPkScriptPath,
+			FpEncKeys:                       fpsEncKeys,
+		})
 		if err != nil {
-			ce.logger.Error("failed to slash unbonding signature", zap.Error(err))
+			ce.logger.Error("failed to sign transactions", zap.Error(err))
 			continue
 		}
 
-		// record metrics
-		finishSignTime := time.Now()
-		metricsTimeKeeper.SetPreviousSignFinish(&finishSignTime)
-		timedSignDelegationLag.Observe(time.Since(startSignTime).Seconds())
-
-		// 8. collect covenant sigs
 		covenantSigs = append(covenantSigs, &types.CovenantSigs{
 			PublicKey:             ce.pk,
 			StakingTxHash:         stakingTx.TxHash(),
-			SlashingSigs:          slashSigs,
-			UnbondingSig:          unbondingSig,
-			SlashingUnbondingSigs: slashUnbondingSigs,
+			SlashingSigs:          resp.SlashSigs,
+			UnbondingSig:          resp.UnbondingSig,
+			SlashingUnbondingSigs: resp.SlashUnbondingSigs,
 		})
 	}
 
-	// 9. submit covenant sigs
-	res, err := ce.cc.SubmitCovenantSigs(covenantSigs)
+	// 10. submit covenant sigs
+	res, err := ce.cc.SubmitCovenantSigs(SortCovenantSigs(covenantSigs))
```

Comment:
hmm why the sorting addtition?  It seems natural to send messages in order of the delegations ? (even though order in which return delegations from Babylon is essentially random 😅 )

Reply:
Good point, I had the sort when it was using the map iteration to build the covenant signatures and since it was a map there was no order than the test `FuzzAddCovenantSig` was failing, but now that we continue the order from the BTC delegations received, there is no need for sorting 
I will remove it

Reply:
removed at https://github.com/babylonlabs-io/covenant-emulator/pull/20/commits/9617ca43d9cbf1f8328fd66aa3aa7c55dc44b1f0